### ISSUE TITLE
hlint-3.0: "Avoid lambda" suggestion incorrectly eta reduces expression

### ISSUE THREAD

**Issue by RyanGlScott** [ndmitchell/hlint#965]
Here's a program:

```hs
module Bug where

data Foo = MkFoo String Int String

f :: [Foo]
f = map (\s -> MkFoo s 0 s) ["a","b","c"]
```

If you run this through `hlint-3.0`, it will provide this suggestion:

```
$ hlint --version
HLint v3.0.2, (C) Neil Mitchell 2006-2020
$ hlint Bug.hs
Bug.hs:6:10-26: Suggestion: Avoid lambda
Found:
  \ s -> MkFoo s 0 s
Perhaps:
  MkFoo s 0

1 hint
```

This is not a valid suggestion, however, since that will fail to compile:

```
Bug.hs:6:16: error: Variable not in scope: s :: String
  |
6 | f = map (MkFoo s 0) ["a","b","c"]
  |                ^
```

This appears to be an `hlint-3.0`–specific bug, since this does not occur with `hlint-2.2.11`.

**Comment by ndmitchell** [comment]
Thanks for the report. I've fixed it and will make a release shortly.

**Comment by ndmitchell** [comment]
3.0.3 out with the fix.
